clients/roscpp/src/libros/init.cpp

```diff
@@ -595,7 +595,8 @@ void shutdown()
   {
     g_internal_queue_thread.join();
   }
-
+  ros::console::deregister_appender(g_rosout_appender);
+  delete g_rosout_appender;
   g_rosout_appender = 0;
```

Comment:
Maybe try wrapping these three lines in `if (g_rosout_appender) { ... }`.

Reply:
Well, it won't work since `g_rosout_appender` is a valid pointer. However, in rosconsole side `g_log4cxx_appender` remains NULL.

Reply:
I'm preparing to submit a PR for non-NULL check of `g_log4cxx_appender` in rosconsole repository. However, it is a little bit awkward while `g_rosout_appender` is not NULL and `g_log4cxx_appender` is NULL. Is it the correct behavior?

Reply:
When I check again, there are two possible appenders. One is glog and the other is log4cxx. I believe, wrong type's `deregister_appender` function is being called.

Reply:
From reading this file `g_rosout_appender` is only set when also being passed to `register_appender` (within the `start()` function). Therefore I think it needs a detailed stacktrace when / why that is the case.

Reply:
Hi again Dirk,

I have inspected `shutdown` and `deregister_appender` functions in `rosconsole`. Following are the functions and my diagnosis:

```
void shutdown()
{
  if(g_log4cxx_appender)
  {
    const log4cxx::LoggerPtr& logger = log4cxx::Logger::getLogger(ROSCONSOLE_ROOT_LOGGER_NAME);
    logger->removeAppender(g_log4cxx_appender);
    g_log4cxx_appender = 0; // **Nullification**
  }
  // reset this so that the logger doesn't get crashily destroyed
  // again during global destruction.  
  //
  // See https://code.ros.org/trac/ros/ticket/3271
  //
  log4cxx::Logger::getRootLogger()->getLoggerRepository()->shutdown();
}
```
,
```
void deregister_appender(LogAppender* appender){
  if(g_log4cxx_appender->getAppender() == appender) //** Trying NULL->getAppender() --> SEGFAULT **/
  {
    const log4cxx::LoggerPtr& logger = log4cxx::Logger::getLogger(ROSCONSOLE_ROOT_LOGGER_NAME);
    logger->removeAppender(g_log4cxx_appender);
    delete g_log4cxx_appender;
    g_log4cxx_appender = 0;
  }
}
```

In my PR, I was trying to call `deregister_appender` after `shutdown`, which directly results in a Segmentation Fault that makes the tests fail. In fact, I have implemented this feature in [here](https://github.com/ros/rosconsole/pull/17) and [here](https://github.com/ros/rosconsole/pull/23). Since all the tests had passed, I thought there was no problem. However `shutdown()` and `deregister_appender()` results in segmentation fault in any context when used subsequently. 
Furthermore, ` logger->removeAppender(g_log4cxx_appender);` already `delete`s `g_log4cxx_appender`. Therefore `deregister_appender` is problematic since the beginning.

So, we don't need to intervene in `rosconsole` level. What is necessary is to just delete `g_rosout_appender` in `init.cpp` and then nullify it according to conventions.

Reply:
Finally, the result of `valgrind --leak-check=full ./devel_isolated/test_roscpp/lib/test_roscpp/test_roscpp-left_right` is below:

```
==5754== Memcheck, a memory error detector
==5754== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==5754== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==5754== Command: ./devel_isolated/test_roscpp/lib/test_roscpp/test_roscpp-left_right
==5754== 
[ INFO] [1555272531.408215746]: ping!
[ INFO] [1555272531.886234315]: ping!
[ INFO] [1555272532.383400126]: ping!y los
[ INFO] [1555272532.883404174]: ping!
[ INFO] [1555272533.383366457]: ping!
==5754== 
==5754== HEAP SUMMARY:
==5754==     in use at exit: 3,457 bytes in 44 blocks
==5754==   total heap usage: 2,016 allocs, 1,972 frees, 296,086 bytes allocated
==5754== 
==5754== LEAK SUMMARY:
==5754==    definitely lost: 0 bytes in 0 blocks
==5754==    indirectly lost: 0 bytes in 0 blocks
==5754==      possibly lost: 0 bytes in 0 blocks
==5754==    still reachable: 3,457 bytes in 44 blocks
==5754==         suppressed: 0 bytes in 0 blocks
==5754== Reachable blocks (those to which a pointer was found) are not shown.
==5754== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==5754== 
==5754== For counts of detected and suppressed errors, rerun with: -v
==5754== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
You can check current repository for the exact same test node. There will be a number of possibly lost bytes. This version produces the above result.